### PR TITLE
Added prelude configuration for project.

### PR DESCRIPTION
**Pull request by orizi**
**Stack**:
- #4311
- #4306
- #4305
- #4304 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)

### CODE REVIEW

**Review by mkaput — CHANGES_REQUESTED** [review]
> Reviewed 2 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: 2 of 5 files reviewed, 2 unresolved discussions (waiting on @orizi and @yuvalsw)

___
*[`corelib/src/prelude_v2023_10.cairo` line 1 at r1](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhkPolW9BOvuBUIGuCx:-NhkPolW9BOvuBUIGuCy:b-3urhf7) ([raw file](https://github.com/starkware-libs/cairo/blob/c9ac0a546b2480664757858738cde5ce682cf683/corelib/src/prelude_v2023_10.cairo#L1)):*
> ```Cairo
> use core::{RangeCheck, SegmentArena, assert, bool, drop, dup, felt252, panic_with_felt252, usize};
> ```

This file seems to have slipped from higher PRs, isn't it?
___
*[`crates/cairo-lang-filesystem/src/db.rs` line 44 at r1](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhkQwbR2QblYHHmsBC1:-NhkQwbS4-6q6TN6DJZP:bu51zy6) ([raw file](https://github.com/starkware-libs/cairo/blob/c9ac0a546b2480664757858738cde5ce682cf683/crates/cairo-lang-filesystem/src/db.rs#L44)):*
> ```XML
>     #[default]
>     V0,
> }
> ```

so that we'll drop the `V` prefix. wdyt?



_Suggestion:_
```XML
#[derive(Clone, Debug, Default, Hash, PartialEq, Eq, Serialize, Deserialize)]
pub enum CompatibilityVersion {
    /// The base version for compatibility.
    #[default]
    #[serde(name = "0")}
    V0,
}
```

**Review by mkaput — COMMENTED** [review]
> Reviewed 3 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 2 unresolved discussions (waiting on @orizi and @yuvalsw)

**Review by orizi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: 1 of 5 files reviewed, 1 unresolved discussion (waiting on @mkaput and @yuvalsw)

___
*[`corelib/src/prelude_v2023_10.cairo` line 1 at r1](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhkPolW9BOvuBUIGuCx:-NhlNnYiAEBALPgrjvzz:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/c9ac0a546b2480664757858738cde5ce682cf683/corelib/src/prelude_v2023_10.cairo#L1)):*
<details><summary><i>Previously, mkaput (Marek Kaput) wrote…</i></summary><blockquote>

This file seems to have slipped from higher PRs, isn't it?
</blockquote></details>

Done.

**Review by yuvalsw — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 7 unresolved discussions (waiting on @mkaput and @orizi)

___
*[`crates/cairo-lang-compiler/src/project.rs` line 76 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvQD-CCxQ_mA7yUuCr:-NhvQD-CCxQ_mA7yUuCs:bbyzij3) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-compiler/src/project.rs#L76)):*
> ```XML
>             .get(&crate_name)
>             .unwrap_or(&crate_config.global)
>             .compatibility;
> ```

This is better to be inside a config's method.



_Code quote:_
```XML
        let compatibility_version = crate_config
            .crate_override
            .get(&crate_name)
            .unwrap_or(&crate_config.global)
            .compatibility;
```
___
*[`crates/cairo-lang-project/src/lib.rs` line 38 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvOM450m6f-VBZnqZM:-NhvOM450m6f-VBZnqZN:b-ydmypz) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/lib.rs#L38)):*
> ```XML
>     pub crate_roots: OrderedHashMap<SmolStr, PathBuf>,
>     #[serde(default)]
>     pub crate_config: CratesConfigContent,
> ```

I really don't like the "content" here and in the other new structs. I just don't get how it's related.
Also suggest to rename the field name to just `config`.



_Suggestion:_
```XML
pub config: CratesConfig,
```
___
*[`crates/cairo-lang-project/src/lib.rs` line 43 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvNK5MABYdIr6UtTNS:-NhvNK5MABYdIr6UtTNT:be8wshv) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/lib.rs#L43)):*
> ```XML
> /// Additional configurations for all crates.
> #[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
> pub struct CratesConfigContent {
> ```

Please rename to something less similar to CrateConfigContent.
e.g.: AllCratesConfigContent

Another option is to call it PerCrateConfigContent, make the fields private and have a get_config_content(crate_id) method.
___
*[`crates/cairo-lang-project/src/lib.rs` line 56 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvN35m8IFkAEuOqr_J:-NhvN35m8IFkAEuOqr_K:bgyky8n) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/lib.rs#L56)):*
> ```XML
> pub struct CrateConfigContent {
>     /// The compatibility version.
>     pub compatibility: CompatibilityVersion,
> ```

does it have to be compatible with only a single version?
___
*[`crates/cairo-lang-project/src/test.rs` line 10 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvPTDm7By6Dmu0GQDQ:-NhvPTDm7By6Dmu0GQDR:b-jo30ta) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/test.rs#L10)):*
> ```XML
>         crate_roots: [("crate_name".into(), "dir".into())].into_iter().collect(),
>         crate_config: CratesConfigContent {
>             global: CrateConfigContent { compatibility: Default::default() },
> ```

I guess you did that becuase you didn't want to derive Default for CrateConfigContent for future compatibility, right?



_Code quote:_
```XML
compatibility: Default::default()
```
___
*[`crates/cairo-lang-project/src/test.rs` line 22 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvPvc87sVjqoHqzMnE:-NhvPvc87sVjqoHqzMnF:brw4mjy) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/test.rs#L22)):*
> ```XML
>     assert_eq!(
>         serialized,
>         indoc! { r#"
> ```

Can you please have 3 crates here, 2 of them overridden (preferably differently)?

**Review by orizi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: 2 of 5 files reviewed, 7 unresolved discussions (waiting on @mkaput and @yuvalsw)

___
*[`crates/cairo-lang-compiler/src/project.rs` line 76 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvQD-CCxQ_mA7yUuCr:-Nhvs2vm6MzEBfs81LRb:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-compiler/src/project.rs#L76)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

This is better to be inside a config's method.
</blockquote></details>

Done.
___
*[`crates/cairo-lang-project/src/lib.rs` line 38 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvOM450m6f-VBZnqZM:-NhvsIXk6JYzLn71nBbv:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/lib.rs#L38)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

I really don't like the "content" here and in the other new structs. I just don't get how it's related.
Also suggest to rename the field name to just `config`.
</blockquote></details>

Done.
___
*[`crates/cairo-lang-project/src/lib.rs` line 43 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvNK5MABYdIr6UtTNS:-Nhvq0e787gT3uFVDPrd:b-pglp51) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/lib.rs#L43)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

Please rename to something less similar to CrateConfigContent.
e.g.: AllCratesConfigContent

Another option is to call it PerCrateConfigContent, make the fields private and have a get_config_content(crate_id) method.
</blockquote></details>

The main idea of this is to be serdeble, so only the first is a viable solution.
___
*[`crates/cairo-lang-project/src/lib.rs` line 56 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvN35m8IFkAEuOqr_J:-NhvqAp-79S0u3sfcOUq:btmj6w3) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/lib.rs#L56)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

does it have to be compatible with only a single version?
</blockquote></details>

I don't understand the question.
___
*[`crates/cairo-lang-project/src/test.rs` line 10 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvPTDm7By6Dmu0GQDQ:-NhvsT903IWnGCjRjeG8:b-e98exh) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/test.rs#L10)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

I guess you did that becuase you didn't want to derive Default for CrateConfigContent for future compatibility, right?
</blockquote></details>

basically - also - not a huge fan of adding `Default` impl just for tests.
___
*[`crates/cairo-lang-project/src/test.rs` line 22 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvPvc87sVjqoHqzMnE:-NhvsjwYBA4ToC8V2DhA:b-y5kzup) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/test.rs#L22)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

Can you please have 3 crates here, 2 of them overridden (preferably differently)?
</blockquote></details>

Only one possible configuration for now.

**Review by mkaput — APPROVED** [review]
> Reviewed 2 of 4 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 6 unresolved discussions (waiting on @yuvalsw)

**Review by yuvalsw — APPROVED** [review]
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 2 unresolved discussions (waiting on @orizi)

___
*[`crates/cairo-lang-project/src/lib.rs` line 56 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvN35m8IFkAEuOqr_J:-NhzuUnb63JPVnoaaz7A:bg9wcxt) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/lib.rs#L56)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

I don't understand the question.
</blockquote></details>

I now understand my confusion, and it made my opinion stronger about not liking the name CompatibilityVersion.
This way, it seems like it denotes which version/versions this crate is compatible with (that is, something like can use crates of this version). This is why I asked if it can be multiple versions. I get that this is the (only) version of the configuration that applies to this crate. Thus, I think that CompilerConfigVersion is much clearer.
___
*[`crates/cairo-lang-project/src/lib.rs` line 39 at r3](https://reviewable.io/reviews/starkware-libs/cairo/4304#-Nhzu8Ok3wUEcMJBfx_h:-Nhzu8Ol9lYggNQzp-jc:b-3qnc53) ([raw file](https://github.com/starkware-libs/cairo/blob/02215945d745d6bcba1ea272416bc111768f23af/crates/cairo-lang-project/src/lib.rs#L39)):*
> ```XML
>     pub crate_roots: OrderedHashMap<SmolStr, PathBuf>,
>     #[serde(default)]
>     pub crate_config: AllCratesConfig,
> ```

suggestion



_Suggestion:_
```XML
b config: AllCratesConfig,
```

**Review by mkaput — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 2 unresolved discussions (waiting on @orizi)

**Review by orizi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: 1 of 5 files reviewed, 2 unresolved discussions (waiting on @mkaput and @yuvalsw)

___
*[`crates/cairo-lang-project/src/lib.rs` line 56 at r2](https://reviewable.io/reviews/starkware-libs/cairo/4304#-NhvN35m8IFkAEuOqr_J:-Ni-rSiu-wi4vfj7wF9W:b-rwj14c) ([raw file](https://github.com/starkware-libs/cairo/blob/1e5aa9b01e9ba08de73a329de59a81cac168cc36/crates/cairo-lang-project/src/lib.rs#L56)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

I now understand my confusion, and it made my opinion stronger about not liking the name CompatibilityVersion.
This way, it seems like it denotes which version/versions this crate is compatible with (that is, something like can use crates of this version). This is why I asked if it can be multiple versions. I get that this is the (only) version of the configuration that applies to this crate. Thus, I think that CompilerConfigVersion is much clearer.
</blockquote></details>

changed to edition, i liked @mkaput idea matching rust.
___
*[`crates/cairo-lang-project/src/lib.rs` line 39 at r3](https://reviewable.io/reviews/starkware-libs/cairo/4304#-Nhzu8Ok3wUEcMJBfx_h:-Ni-rD6_3PStO-gpJoXw:byno9vw) ([raw file](https://github.com/starkware-libs/cairo/blob/02215945d745d6bcba1ea272416bc111768f23af/crates/cairo-lang-project/src/lib.rs#L39)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

suggestion
</blockquote></details>

note the data here is the names for the toml.
config here seems weird for me.

**Review by mkaput — APPROVED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 2 unresolved discussions (waiting on @yuvalsw)

**Review by yuvalsw — APPROVED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 1 unresolved discussion (waiting on @orizi)

___
*[`crates/cairo-lang-project/src/lib.rs` line 39 at r3](https://reviewable.io/reviews/starkware-libs/cairo/4304#-Nhzu8Ok3wUEcMJBfx_h:-Ni46pKu-kVqQzgtGyXQ:b-7xghl3) ([raw file](https://github.com/starkware-libs/cairo/blob/02215945d745d6bcba1ea272416bc111768f23af/crates/cairo-lang-project/src/lib.rs#L39)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

note the data here is the names for the toml.
config here seems weird for me.
</blockquote></details>

It's of type AllCratesConfig. At least change to crates_config (plural for crate).

**Review by mkaput — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 1 unresolved discussion (waiting on @orizi)

**Review by yuvalsw — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: all files reviewed, 2 unresolved discussions (waiting on @orizi)

___
*[`crates/cairo-lang-project/src/lib.rs` line 49 at r5](https://reviewable.io/reviews/starkware-libs/cairo/4304#-Ni95W_2AVoqv4yuBAy2:-Ni95W_2AVoqv4yuBAy3:btclpha) ([raw file](https://github.com/starkware-libs/cairo/blob/8bf4bd23140a15169a8cc1f27c0248819ca108c1/crates/cairo-lang-project/src/lib.rs#L49)):*
> ```XML
>     /// The configuration for non overriden crates.
>     #[serde(default)]
>     pub all: SingleCrateConfig,
> ```

global was actually better IMO. "all" is worse, as it's not really all...
If you don't like global I can suggest default/base.
___
*[`crates/cairo-lang-project/src/lib.rs` line 53 at r5](https://reviewable.io/reviews/starkware-libs/cairo/4304#-Ni95MhS6jepNwDkyyjv:-Ni95MhS6jepNwDkyyjw:b-2hftwt) ([raw file](https://github.com/starkware-libs/cairo/blob/8bf4bd23140a15169a8cc1f27c0248819ca108c1/crates/cairo-lang-project/src/lib.rs#L53)):*
> ```XML
>     #[serde(default)]
>     #[serde(rename = "override")]
>     pub override_: OrderedHashMap<SmolStr, SingleCrateConfig>,
> ```

overrides? override_map?



_Code quote:_
```XML
override_
```

**Review by orizi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304)* status: 3 of 5 files reviewed, 2 unresolved discussions (waiting on @yuvalsw)

___
*[`crates/cairo-lang-project/src/lib.rs` line 49 at r5](https://reviewable.io/reviews/starkware-libs/cairo/4304#-Ni95W_2AVoqv4yuBAy2:-Ni9Ia615ufkJTbVFiZC:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/8bf4bd23140a15169a8cc1f27c0248819ca108c1/crates/cairo-lang-project/src/lib.rs#L49)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

global was actually better IMO. "all" is worse, as it's not really all...
If you don't like global I can suggest default/base.
</blockquote></details>

Done.
___
*[`crates/cairo-lang-project/src/lib.rs` line 53 at r5](https://reviewable.io/reviews/starkware-libs/cairo/4304#-Ni95MhS6jepNwDkyyjv:-Ni9IavM1jK3o6V1cE9T:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/8bf4bd23140a15169a8cc1f27c0248819ca108c1/crates/cairo-lang-project/src/lib.rs#L53)):*
<details><summary><i>Previously, yuvalsw wrote…</i></summary><blockquote>

overrides? override_map?
</blockquote></details>

Done.

**Review by yuvalsw — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/4304#-:-NiADjY82kXRKnP0wo4e:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @orizi)
